FreeRTOS-Plus/Demo/FreeRTOS-IoT-Libraries-LTS-Beta2/mqtt/mqtt_multitask/DemoTasks/MultitaskMQTTExample.c

```diff
@@ -0,0 +1,1457 @@
+            xStatus = MQTT_Publish( &globalMqttContext, pxPublishInfo, usPacketId );
+            pxCommand->pContext->returnStatus = xStatus;
+
+            /* Add to pending ack list, or call callback if QoS 0. */
+            xAddAckToList = ( pxPublishInfo->qos != MQTTQoS0 ) && ( xStatus == MQTTSuccess );
+            break;
+
+        /* TODO: Option to subscribe/unsubscribe without sending a packet,
+         * e.g. for Shadow topics. */
+        case SUBSCRIBE:
+        case UNSUBSCRIBE:
+            assert( pxCommand->pContext != NULL );
+            pxSubscribeInfo = pxCommand->pContext->pSubscribeInfo;
+            assert( pxSubscribeInfo != NULL );
+            assert( pxSubscribeInfo->pTopicFilter != NULL );
+            usPacketId = MQTT_GetPacketId( &globalMqttContext );
+
+            if( pxCommand->commandType == SUBSCRIBE )
+            {
+                LogInfo( ( "Subscribing to %.*s",
+                           pxSubscribeInfo->topicFilterLength,
+                           pxSubscribeInfo->pTopicFilter ) );
+                xStatus = MQTT_Subscribe( &globalMqttContext,
+                                          pxSubscribeInfo,
+                                          pxCommand->pContext->subscriptionCount,
+                                          usPacketId );
+            }
+            else
+            {
+                LogInfo( ( "Unsubscribing from %.*s", pxSubscribeInfo->topicFilterLength, pxSubscribeInfo->pTopicFilter ) );
+                xStatus = MQTT_Unsubscribe( &globalMqttContext,
+                                            pxSubscribeInfo,
+                                            pxCommand->pContext->subscriptionCount,
+                                            usPacketId );
+            }
+
+            pxCommand->pContext->returnStatus = xStatus;
+            xAddAckToList = ( xStatus == MQTTSuccess );
+            break;
+
+        case PING:
+            xStatus = MQTT_Ping( &globalMqttContext );
+
+            if( pxCommand->pContext != NULL )
+            {
+                pxCommand->pContext->returnStatus = xStatus;
+            }
+
+            break;
+
+        case DISCONNECT:
+            xStatus = MQTT_Disconnect( &globalMqttContext );
+
+            if( pxCommand->pContext != NULL )
+            {
+                pxCommand->pContext->returnStatus = xStatus;
+            }
+
+            break;
+
+        case CONNECT:
+            /* TODO: Reconnect. */
+            LogInfo( ( "Processed Connect Command" ) );
+
+        default:
+            break;
+    }
+
+    if( xAddAckToList )
+    {
+        prvAddAck( usPacketId, pxCommand->pContext, pxCommand->callback );
+    }
+    else
+    {
+        if( pxCommand->callback != NULL )
+        {
+            pxCommand->callback( pxCommand->pContext );
+        }
+    }
+
+    return xStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool matchEndWildcards( const char * pTopicFilter,
```

Comment:
for all the topic matching, can we use the API that was added to the MQTT lib recently?

Reply:
It was merged in the CSDK 10 minutes before this comment, please give me some more time to update this PR lol